libvast/src/address.cpp

```diff
@@ -16,13 +16,103 @@
 
 #include <arpa/inet.h>
 #include <netinet/in.h>
+#include <openssl/crypto.h>
+#include <openssl/err.h>
+#include <openssl/evp.h>
+#include <openssl/ossl_typ.h>
 #include <sys/socket.h>
 
 #include <cstdlib>
 #include <utility>
 
 namespace vast {
 
+namespace {
+inline uint32_t bitmask32(size_t bottom_bits) {
+  return bottom_bits >= 32 ? 0xffffffff : ((uint32_t{1} << bottom_bits) - 1);
+}
+
+class AddressEncryptor {
+public:
+  AddressEncryptor(const std::array<address::byte_type, 32>& key) {
+    ctx_ = EVP_CIPHER_CTX_new();
+    EVP_CIPHER_CTX_init(ctx_);
+    OpenSSL_add_all_ciphers();
+    cipher_ = EVP_get_cipherbyname("aes-128-ecb");
+    block_size_ = EVP_CIPHER_block_size(cipher_);
+    pad_ = std::vector<address::byte_type>(block_size_);
+    auto pad_out_len = 0;
+    EVP_CipherInit_ex(ctx_, cipher_, nullptr, key.data(), nullptr, 1);
+    // use second 16-byte half of key for padding
+    EVP_CipherUpdate(ctx_, pad_.data(), &pad_out_len, key.data() + block_size_,
+                     block_size_);
+  }
+
+  AddressEncryptor(const AddressEncryptor&) = delete;
+
+  ~AddressEncryptor() {
+    EVP_CIPHER_CTX_cleanup(ctx_);
+    EVP_CIPHER_CTX_free(ctx_);
+  }
+
+  auto encrypt(address::byte_array bytes, size_t byte_offset) {
```

Comment:
Would it make sense to make this the call operator, so you don't `encryptor.encrypt(...)`, but rather `encryptor(...)`? Just reads a lot nicer to me.

(If you disagree just resolve, I have no particularly strong opinion on this.)

Reply:
Good idea. Implemented in 8f1d2581d